### PR TITLE
Full Averaging Incorporation

### PR DESCRIPTION
**Pull request by nbollis**
Eliminating the SpectralAveraging nuget package and creating a project within MzLib

I have successfully packaged this and loaded the SpectralAveraging project in MetaMorpheus

The bulk of the code can be found in the Algorithms folder. These static classes are responsible for processing the data with a case structure to allow for easy expansion in the future. Each algorithm has an accompanying enum representing the different types of algorithms. 

MzBinning is the main workhorse of this PR. This SpectralAveraging method will normalize spetra, determine their weight for the weighted averaging of peaks, and sort the peaks into bins. Each bin will then have its outliers rejected and be averaged. The averaged bins will then be recapitulated into x and y arrays before being returned. This results in most of the code in this PR being executed within the method below

image

Below is a brief explanation of the structure for each static algorithm class

SpectraFileAveraging 
- Takes a large collection of scans, bundles them up as specified by the enum, averages each bundle, and finally recapitulates a MsDataFile
- Types 
   - Average All - returns one averaged spectra
   - Average Every N Scans - returns file of averaged spectral
   - Average DDA Scans - averages MS1 spectra and leaves MS2 untouched

Spectral Averaging 
- Averages a set of scans
- Has Extension methods for collections of MsDataScans and MzSpectrum
- Types
   - MzBinning - Splits the mz axis into bins and averages each bin

Spectral Normalization
- Normalizes a set of spectra
- Has Extension methods for single object and collections of MsDataScans and MzSpectrum
- Types
   - AbsoluteToTic - divide each intensity by its tic value
   - RelativeToTics - divide each intensity by its tic value and multiply by average tic value

Spectral Weighting
- Determine weight to be given to each spectra when averaging. 
- Types
   - Weight Evenly
   - Weight By Tic Value 
   - Weight By Noise Estimation - coming soon

Outlier Rejection
- Reject outliers from each bin of peaks
- Types
   -  MinMax Clipping - remove greatest and lowest value
   - Percentile Clipping- remove values x percentile from median value
   - Sigma Clipping - iteratively remove values that fall outside of the central value by the defined StandardDeviation amount
   - Winsorized Sigma Clipping - iteratively remove values that fall outside of the central value by the defined StandardDeviation amount with the values of the median * that standard deviation amount
   - Averaged Sigma Clipping - iteratively remove values that fall outside of a calculated deviation based upon the median of the values
   - Below Threshold Rejection - eliminates bins that have too few peaks when compared to the number of scans being averaged. 

Averaged Spectra Outputter
- Outputs the averages spectra
- Types
    - mzML
    - text - folder of txt files, each containing x and y of an averaged spectra

### CODE REVIEW

**Review by Alexander-Sol — CHANGES_REQUESTED** [review]
Test and TestFlashLFQ .csproj files contain an outdated reference to MzLibSpectralAveraging/SpectralAveraging.csproj that cause build failures

**Review by Alexander-Sol — COMMENTED** [review]
I found AveragedSpectraOutputter to be unintuitive. AveragedSpectraWriter would be more idiomatic.

**Review by Alexander-Sol — CHANGES_REQUESTED** [review]
Looks great, beautifully encapsulated, extremely readable code. Only a couple of minor issues: "Outputter" instead of "Writer" and creating new MS2 scans instead of modifying in place. (Both issued more clearly defined in individual comments.)

**Review by Alexander-Sol — APPROVED** [review]

File: mzLib/SpectralAveraging/Algorithms/SpectraFileAveraging.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            var newNativeID =
+                averagedScan.NativeId.Replace(averagedScan.NativeId.Split("=").Last(), scanNumberIndex.ToString());
+            averagedScan.SetNativeID(newNativeID);
+            averagedScans.Add(averagedScan);
+            var precursorScanIndex = scanNumberIndex;
+            scanNumberIndex++;
+
+            foreach (var scan in ms2ScansFromAveragedScans)
+            {
+                MsDataScan newScan = new(scan.MassSpectrum, scanNumberIndex, scan.MsnOrder, scan.IsCentroid,
```
[the remark this answers is not in the document]
Comment (nbollis) [reply mzLib/SpectralAveraging/Algorithms/SpectraFileAveraging.cs]: This has been fixed in the recent commits

**Comment by nbollis** [comment]
> Looks great, beautifully encapsulated, extremely readable code. Only a couple of minor issues: "Outputter" instead of "Writer" and creating new MS2 scans instead of modifying in place. (Both issued more clearly defined in individual comments.)

Both of these have been addressed and changed

**Review by acesnik — APPROVED** [review]
I haven't done a deep review, but this looks well tested and well integrated.

**Review by Alexander-Sol — APPROVED** [review]

**Pull request merged**